src/mame/drivers/firebeat.cpp

```diff
@@ -176,445 +176,522 @@ struct IBUTTON
 	required_shared_ptr<uint32_t> m_work_ram;
-	optional_device<pc16552_device> m_duart_midi;
-	required_device<pc16552_device> m_duart_com;
-	optional_device_array<midi_keyboard_device, 2> m_kbd;
 	required_device<ata_interface_device> m_ata;
-	optional_device_array<k057714_device, 2> m_gcu;
-	optional_device<cy7c131_device> m_dpram;
-	optional_device<ata_interface_device> m_spuata;
-	optional_shared_ptr<uint16_t> m_waveram;
+	required_device<k057714_device> m_gcu;
 
-	uint8_t m_extend_board_irq_enable;
-	uint8_t m_extend_board_irq_active;
-//  emu_timer *m_keyboard_timer;
-	int m_layer;
-	int m_cab_data_ptr;
-	const int * m_cur_cab_data;
-//  int m_keyboard_state[2];
-	IBUTTON m_ibutton;
-	int m_ibutton_state;
-	int m_ibutton_read_subkey_ptr;
-	uint8_t m_ibutton_subkey_data[0x40];
+private:
+	uint32_t cabinet_r(offs_t offset, uint32_t mem_mask = ~0);
 
-	uint32_t m_spu_ata_dma;
-	int m_spu_ata_dmarq;
-	uint32_t m_wave_bank;
+	void set_ibutton(uint8_t *data);
+	int ibutton_w(uint8_t data);
+	void security_w(uint8_t data);
 
-	void firebeat_spu_base(machine_config &config);
+	uint32_t extend_board_irq_r(offs_t offset, uint32_t mem_mask = ~0);
+	void extend_board_irq_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
 
-	uint32_t screen_update_firebeat_0(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
-	uint32_t screen_update_firebeat_1(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
-	INTERRUPT_GEN_MEMBER(firebeat_interrupt);
 	uint32_t input_r(offs_t offset, uint32_t mem_mask = ~0);
-	uint32_t sensor_r(offs_t offset);
-	DECLARE_WRITE_LINE_MEMBER(ata_interrupt);
+
 	uint32_t ata_command_r(offs_t offset, uint32_t mem_mask = ~0);
 	void ata_command_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+
 	uint32_t ata_control_r(offs_t offset, uint32_t mem_mask = ~0);
 	void ata_control_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+
 //  uint32_t comm_uart_r(offs_t offset, uint32_t mem_mask = ~ 0);
 //  void comm_uart_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	uint32_t cabinet_r(offs_t offset, uint32_t mem_mask = ~0);
-	uint32_t keyboard_wheel_r(offs_t offset);
-	uint8_t midi_uart_r(offs_t offset);
-	void midi_uart_w(offs_t offset, uint8_t data);
-	uint32_t extend_board_irq_r(offs_t offset, uint32_t mem_mask = ~0);
-	void extend_board_irq_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void lamp_output_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void lamp_output_kbm_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void lamp_output_ppp_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void lamp_output2_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void lamp_output2_ppp_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void lamp_output3_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void lamp_output3_ppp_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+
+	IBUTTON m_ibutton;
```

Comment:
I was also tempted to do something about this since it stood out as something that didn't quite belong in Firebeat directly. I'll take a look at it when I have some time, maybe include it in my next PR.